### PR TITLE
More setup env changes

### PR DESCRIPTION
**Pull request by kekiefer**
These setup-env changes prepare for multiple distro support, where the distros can have different template bblayers or configuration options.

The contents of the .templateconf are changed to interpolate the `DISTRO` variable and choose a subdirectory that contains the template files specific to this `DISTRO`. When re-sourcing an existing build directory (and thereby not re-specifying the `DISTRO`), the template.conf from the build directory is used so the interpolation does not need to happen.

I'm now printing the conf-notes.txt from that config layer, so the user should be able to see at a glance how the build directory was originally configured by observing different messages in the case of multiple distros.

I've also added a diff that shows a warning if there are changes in the bblayers.conf vs. the bblayers.conf.sample that the build directory was originally configured with.

The place this diff falls a bit short is for what I really wanted, to see if the `DISTRO` value is changed manually in local.conf after the build directory has been configured, and the layer set needs to be updated. If I had a good way to figure out the actual value of `DISTRO`, I could look up the corresponding template directory to use in the bblayers diff. I can't think of a good way to do this except by some really complicated parsing of the local.conf that I'd prefer to avoid. Open to suggestions here.

### CODE REVIEW

File: setup-env [hunk 1]
```diff
@@ -191,9 +176,6 @@ else
     echo "Your configuration files at $BUILDDIR have not been touched."
 fi
 
-# sets TEMPLATECONF
-. $TDROOT/.templateconf
```
Comment (madisongh) [inline setup-env:195]: Did this get dropped for a reason?  Shouldn't it have been moved to just under where `TEMPLATECONF` get set (line 164)?

**Review by madisongh — COMMENTED** [review]
This change breaks my existing workspaces due to the relocation of the templates, but that's OK.

I had one question in-line.

File: setup-env (hunk 1 above)
Comment (kekiefer) [reply setup-env:195]: This same functionality is moved to line 164 (`TEMPLATECONF=$(cat "$BUILDDIR/conf/templateconf.cfg")` it is set from the $BUILDDIR/conf/templateconf.cfg which doesn't require the `DISTRO` variable. It had to be moved up as it's required to cat the conf-notes.txt.

**Comment by kekiefer** [comment]
It's true about the existing workspaces. I could add a hard copy of the bblayers.conf for the main testdistro and symlink to that for the default template-testdistro and that would probably be nicer.

File: setup-env (hunk 1 above)
Comment (madisongh) [reply setup-env:195]: Right, but if `templateconf.cfg` doesn't exist yet (i.e., new build dir) sourcing the `.templateconf` file will set it appropriately. Having both mimics the sequence in `oe-setup-builddir`, which I think is what we want here.

**Comment by madisongh** [comment]
> It's true about the existing workspaces. I could add a hard copy of the bblayers.conf for the main testdistro and symlink to that for the default template-testdistro and that would probably be nicer.

Don't worry about it.  It's a one-time conversion and I'm probably the only person on the planet who is affected.

**Comment by kekiefer** [comment]
Ha, too late! I think it makes sense, since this is the default distro. This way, if someone wants to set TEMPLATECONF manually, the default will be in the standard location.

File: setup-env (hunk 1 above)
Comment (kekiefer) [reply setup-env:195]: `oe-setup-builddir` is run prior to this step, and it is responsible for writing `templateconf.cfg` with the interpolated value it generated from `.templateconf` before we try to use it here.

Comment (madisongh) [reply setup-env:195]: Ah, OK. That's the bit I was missing just looking at the diffs.

**Comment by madisongh** [comment]
> The place this diff falls a bit short is for what I really wanted, to see if the `DISTRO` value is changed manually in local.conf after the build directory has been configured, and the layer set needs to be updated. If I had a good way to figure out the actual value of `DISTRO`, I could look up the corresponding template directory to use in the bblayers diff. I can't think of a good way to do this except by some really complicated parsing of the local.conf that I'd prefer to avoid. Open to suggestions here.

This maybe could be handled through a sanity check function that gets called when bitbake is started.  There's a BBLAYERS_CONF_UPDATE_FUNCS variable that can be set to specify a config file, a couple of variables, and an update function.  As long as we keep the layers-config version number ranges of the different distros in the repo disjoint, we could have a function that flags a mismatch and can handle auto-upgrades of the layers when the distro modifies its layer set.   It would be an extension of what's done in `meta-poky/classes/poky-sanity.bbclass` in the poky repo.

**Comment by madisongh** [comment]
One more nit about these changes, btw - I'm not a big fan of assuming that the output device is a terminal that can handle ANSI color sequences.  Making the coloring conditional would be an improvement... in fact, how about moving that section of commands (with the layer set diffs and such) into a script? There might be other sections there that could be moved as well. That would also help keep the size of setup-env more under control.

**Comment by kekiefer** [comment]
Those update functions might be exactly what I was looking for. Thanks, I'll see if there's an opportunity there, and I'll break that diff stuff out.

**Comment by kekiefer** [comment]
OK, I've made these two updates.

The diff is broken out into the scripts directory, and colorizes by default when output is to a terminal. This can be overridden with a new `--color` option, which takes the same `always`, `never`, `auto` options as normal GNU tools, and should function the same way.

`BBLAYERS_CONF_UPDATE_FUNCS` is a little more limited than I'd have liked. It relies on doing a simple equality comparison in python. At the end of the day, this should enough for the sanity check I want -- now it'll stop you if you changed `DISTRO` and didn't update the bblayers.conf to match. It doesn't actually fix up the layers (or even check their validity), but I think the hint should be enough. You can skip this check by taking the `LCONF_DISTRO` line out in your build directory.

If we want to get fancy, the new `BBLAYERS_CONF_UPDATE_FUNCS` could be used to automatically fix up the layers. I don't love the idea of a script messing with the user's configuration files, and it's certainly much simpler this way. But it's possible.

**Comment by madisongh** [comment]
@kekiefer I merged this manually after reworking the bblayers.conf sanity checks a bit... The version string now includes both the distro name and a version number, and I switched to using new variables for this purpose (rather than the ones defined in OE-Core).  The sanity checker will automatically overwrite the bblayers.conf file when it detects that it needs to -- either a distro change, or a version change.

Verified that it was compatible with my old workspaces, and auto-upgraded just fine.

Thanks

**Comment by kekiefer** [comment]
Thanks for the assist. Looks good.

**Pull request closed without merging**